### ISSUE TITLE
Feature Request: GPU Support

### ISSUE THREAD

**Issue by ZongqiangZhang** [kubernetes/node-problem-detector#833]
This feature request aims to enhance the Node Problem Detector with the ability to monitor GPUs on nodes and detect issues.

Currently NPD does not have direct visibility into GPUs. However, many workloads are GPU accelerated which makes GPU health an important part of node health. e.g. GPUs are widely used in machine learning training and inference. Especially for LLM training which may using tens of thousands of GPU cards. The entire training cluster should be restarted from previous checkpoint if any one of the GPUs in the cluster is gone bad.

This feature request adds the following capabilities:

- GPU device monitoring: NPD will collect GPU device info periodically and look for crashes or errors via nvidia-smi/nvml/dcgm tools.

- GPU device monitoring: NPD will check GPU device info periodically to detect if a GPU is "stuck" (e.g. nvidia-smi command hangs).

- TBD: GPU runtime monitoring: NPD will check for crashes or OOM issues reported in nvidia logs.

Specifically, this feature request includes:
- Code for the gpu_monitor plugin
- A Dockerfile to build an NPD image with GPU support
- Other dependencies

Looking forward to your feedback!

**Comment by zz913922** [comment]
Exactly the same situation, did you resolve this issue already?

**Comment by AllenXu93** [comment]
I think NPD should support different config for different device or runtime;
For example , I have both GPU and none GPU worker node in one cluster, or containerd and docker nodes in one cluster; currently we need to deploy 2 npd daemonset for different type of node;

**Comment by AllenXu93** [comment]
BTW, for GPU, we don't need to install more dependencies, it just add env `NVIDIA_VISIBLE_DEVICES` and `NVIDIA_DRIVER_CAPABILITIES` , and use `nvidia-smi` command to check GPU state

**Comment by wangzhen127** [comment]
Thanks for filing the feature request! I think this totally makes sense. Do you have any more concrete proposal?

/cc @SergeyKanzhelev

**Comment by SergeyKanzhelev** [comment]
yes, accelerators health is an important functionality and would be great to have it in NPD

Need to design it carefully though. There is already some health checking in a device plugin (like https://github.com/NVIDIA/k8s-device-plugin/blob/bf58cc405af03d864b1502f147815d4c2271ab9a/internal/rm/health.go#L39) that we need to work nicely with. Even simple detection of a device plugin health is a good starting point here.

@AllenXu93 @ZongqiangZhang do you want to work on more detailed design? I definitely will be interested to join the effort

**Comment by AllenXu93** [comment]
> yes, accelerators health is an important functionality and would be great to have it in NPD
> 
> Need to design it carefully though. There is already some health checking in a device plugin (like https://github.com/NVIDIA/k8s-device-plugin/blob/bf58cc405af03d864b1502f147815d4c2271ab9a/internal/rm/health.go#L39) that we need to work nicely with. Even simple detection of a device plugin health is a good starting point here.
> 
> @AllenXu93 @ZongqiangZhang do you want to work on more detailed design? I definitely will be interested to join the effort

Of cource. 
In our case, we use `nvidia-smi` to check GPU remapped row pending and failure (accroding to  https://docs.nvidia.com/deploy/a100-gpu-mem-error-mgmt/#error-recovery-and-response-flags ), mark condition of  node. When it occurr, we will create a job that drain node and execute `gpu reset`.  So we need NPD to check GPU health and mark node condition.

**Comment by xuchenCN** [comment]
LGTM + 1

**Comment by AllenXu93** [comment]
您好，邮件已收到，我会尽快给您回复。

**Comment by SergeyKanzhelev** [comment]
The important question here is what NPD will be collecting comparing to the device plugin. Some designing is needed here

**Comment by ravisantoshgudimetla** [comment]
> Need to design it carefully though. There is already some health checking in a device plugin (like https://github.com/NVIDIA/k8s-device-plugin/blob/bf58cc405af03d864b1502f147815d4c2271ab9a/internal/rm/health.go#L39) that we need to work nicely with. Even simple detection of a device plugin health is a good starting point here.

I believe this correct. We are ignoring just a handful of application specific errors, so detection of device plugin is a good starting point.

**Comment by SergeyKanzhelev** [comment]
@ArangoGutierrez suggests that utilizing https://github.com/NVIDIA/NVSentinel would be the best way to go here

**Comment by dims** [comment]
@SergeyKanzhelev @ArangoGutierrez i posted something in slack - https://kubernetes.slack.com/archives/CJA25LM6D/p1761938843472639

```
[@Sergey Kanzhelev](https://kubernetes.slack.com/team/U013U7PFRA5) [@hakman](https://kubernetes.slack.com/team/UQ2FC9405) I was looking at options to integrate nvsentinel ( https://github.com/NVIDIA/NVSentinel/ ) with node-problem-detector. So i dug really deep into how the NPD plugins work currently
https://gist.github.com/dims/6f482d9fe34669c0df8f8f5564474981

I just did not like the options sadly :disappointed: So here's a proposal POC based on how containerd uses grpc to work with external snapshotters.
https://github.com/dims/npd-ext

Highlights:
- the protos are here - https://github.com/dims/npd-ext/tree/main/api/services/external/v1
- there's a clean brand new "external monitor" https://github.com/dims/npd-ext/tree/main/pkg/externalmonitor that does not stomp on any existing code
- fun little gpu-monitor that listens on a socket and uses nvidia-smi to do something useful https://github.com/dims/npd-ext/blob/main/examples/external-plugins/gpu-monitor/main.go

The idea is definitely for nvsentinel (OR anyone else writing a new plugin) to have another option that does not involve recompiling or shelling out to a binary. WDYT?
```

**Comment by dims** [comment]
cc @hakman

**Comment by andrewsykim** [comment]
Is anyone working on this? I would love to be able to integrate NPD signals for GPU failures into Ray / KubeRay ecosystem

**Comment by hakman** [comment]
> Is anyone working on this? I would love to be able to integrate NPD signals for GPU failures into Ray / KubeRay ecosystem

@dims tried something last year, but planets did not align. I think his suggestions had merit.
@SergeyKanzhelev any thoughts on this?

**Comment by dims** [comment]
@hakman i am happy to revive :)

**Comment by dims** [comment]
hola @andrewsykim !!!

**Comment by SergeyKanzhelev** [comment]
We've been discussing with @ArangoGutierrez that one great way for NPD to implement basic GPU monitoring is to hook up into GPU device plugin health endpoint, same as we use for Pod Status: https://github.com/kubernetes/enhancements/issues/4680 This way we will unify on errors representation between Pod Status and Node Conditions.

The issue with this approach is that currently we have a log of DevicePlugin-based systems. So we need an interim solution. If we want to support GPU devices monitoring directly when Device Plugin is used, I would advocate to keep it slim with the idea to transition it to Health API of DRA at some point.

I am not aware of anybody working on it as of now.

**Comment by dims** [comment]
@SergeyKanzhelev not sure if you saw/remember this from last year - https://kubernetes.slack.com/archives/CJA25LM6D/p1761938843472639

**Comment by ganeshkumarashok** [comment]
Excited to see the thread being revived. I'd be happy to share experiences developing GPU NPD on AKS (needed custom health check scripts along with GPU instance-specific behavior) and see if I can contribute here. Having the framework be GPU instance type and vendor neutral, would be great.

**Comment by hakman** [comment]
@ganeshkumarashok this would be definitely welcomed.

**Comment by dims** [comment]
@ganeshkumarashok @hakman @SergeyKanzhelev @andrewsykim - proposing a first step in #1281 (I REALLY don't want to run scripts repeatedly!)
